data_cleaning/lib/cleaning_tools.py

```diff
@@ -163,3 +163,22 @@ def read_dtw_csv(project_key, filename, **kwargs):
     os.write(new_file, data_bytes)
     os.close(new_file)
     return pd.read_csv(tmpfilename, **kwargs)
+
+
+def reorder_columns_and_check(df, new_order):
+    '''Return dataframe with reordered columns, making sure we didn't leave any columns out.'''
+    assert len(new_order) == len(set(new_order)), "Duplicate columns in new_order! Plz fix."
+    assert len(df.columns) == len(set(df.columns)), "Duplicate columns in original dataframe! Plz fix."
+
+    if set(new_order) != set(df.columns):
+        old_cols = list(df.columns)
+        messages = []
+        for c in old_cols:
+            if c not in new_order:
+                messages.append("Column '%s' from the original dataframe is missing in new_order" % c)
+        for c in new_order:
+            if c not in old_cols:
+                messages.append("Column '%s' in new_order does not exist in the original dataframe" % c)
+        assert False, '\n'.join(messages)
```

Comment:
Just making sure I understand what this is doing - if messages isn't empty, the assertion will display those messages, right?

Reply:
Thanks for checking A.

`assert False` will always fail, whether messages is empty or not. But given that all this is within the `if` condition on like 173 (`if set(new_order) != set(df.columns):`), if the sets are unequal, then by definition a column exists in one that doesn't exist in the other. Hence, there will always be at least one message. Even if somehow messages were empty, it would just assert False and have an empty message string, since `'\n'.join([])` is just the empty string.

However, now that I read it, `assert` was the wrong way to go. I'm going to have it raise a CleaningError, a custom Error type I already created just for this file. So thanks for helping me take a closer look -- see code update momentarily.